# FIT5196 Task 1 in Assessment 2
#### Student Name: Yicheng Zhang
#### Student ID: 27699641

Date: 10/05/2018

Version: 1.0

Environment: Python 3.6.5 and Jupyter notebook

Libraries used:
* pandas (for dataframe, included in Anaconda Python 3.6) 
* re (for regular expression, included in Anaconda Python 3.6) 
* numpy (for numpy array, included in Anaconda Python 3.6)
* sys (for system information, included in Anaconda Python 3.6)
* googlemaps(for google map api usage, should be installed before using it)

## 1.  Import libraries 

In [65]:
import pandas as pd
import numpy as np
try:
    import googlemaps #try import googlemaps
except:
    import pip
    !pip install googlemaps
    import googlemaps  # if not exits, install this package
import re
import datetime
import sys

In [66]:
sys.version

'3.6.5 |Anaconda custom (64-bit)| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]'

## 2. Load Data with Pandas

### 2.1. Load CSV File
First step of this assessment is loading the csv file's data from the file, In python,one of the optimized method is using the `pandas` build-in method named `read_csv()`
We can get the data with `dataframe` type.

In [67]:
dataset1 = pd.read_csv('dataset1_with_error.csv')
dataset1.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  not available    permanent  Gregory Martin International  Engineering Jobs   
1  not available    permanent  Gregory Martin International  Engineering Jobs   
2  not available    permanent  Gregory Martin International  Engineering Jobs   
3  not available    permanent  Gregory Martin International  Engineering Jobs   
4  not available    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  
0            24996  cv-library.co.uk  20121103T000000  20121203T000000  
1            30000  cv-library.co.uk  20130108T150000  20130408T150000  
2            30000  cv-library.co.uk  20130726T150000  20130924T150000  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000  
4            24996  cv-library.co.uk  20131025T000000  20131224T000000

In [68]:
print (dataset1.shape)

(25077, 11)


There are total 25077 records in this dataset, and 11 columns. 

### 2.2. Check Data information
There are many methods to check the data information, they can be statistic, or descriptive. The `pandas` package provides us a convenient way to observe data information easily.
- - -
`dataframe.info()` can show `The number of records`, `Columns(Name,Type,Non-null values)`,`memory usage`

It can be found, only the `Id` column has data type `int64` which means, in this columns there are only numeric data.

There are some null value(missing value) in `Company` column

In [69]:
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25077 entries, 0 to 25076
Data columns (total 11 columns):
Id                  25077 non-null int64
Title               25077 non-null object
Location            25077 non-null object
ContractType        25077 non-null object
ContractTime        25077 non-null object
Company             21242 non-null object
Category            25077 non-null object
Salary per annum    25077 non-null object
SourceName          25077 non-null object
OpenDate            25077 non-null object
CloseDate           25077 non-null object
dtypes: int64(1), object(10)
memory usage: 2.1+ MB


`dataframe.describe()` can show the data statistic summary, however, it can only display numeric data(in this case `Id`).

In [70]:
dataset1.describe()

Id
count  2.507700e+04
mean   6.664312e+07
std    5.195261e+06
min    1.261263e+07
25%    6.720830e+07
50%    6.836110e+07
75%    6.871371e+07
max    6.924767e+07

If we want to know other data type's summary, we can use `dataframe.describe(include=['O'])`, which include object type data in python.
- - -
We can get the **count** of data, **unique** record in these data, which record appear most **frequent** in the data.

In [71]:
dataset1.describe(include=['O']) # describe include objects

Title Location   ContractType  \
count                                     25077    25077          25077   
unique                                    25077      482              3   
top     Mechanical Manufacturing Engineer  ****       UK  not available   
freq                                          1     3996          19499   

       ContractTime        Company Category Salary per annum     SourceName  \
count         25077          21242    25077            25077          25077   
unique            3           4879        8             1589             90   
top       permanent  UKStaffsearch  IT Jobs            35004  totaljobs.com   
freq          16194            248     7085             1011           5335   

               OpenDate        CloseDate  
count             25077            25077  
unique             2203             2400  
top     20130127T150000  20130526T120000  
freq                 27               23

There are some interesting discoveries based on the above table:
- - -
* ** Location: ** Most Records in the data have Location with `UK`, There can be relatively high possibility this data come from united kindom.
- - -
* ** Contract: ** There are only 3 types for both contract type and time, it can be found later.
- - -
* ** Company: ** There are some missing values in company data.
- - -
* ** Date: ** The format of date should be considered later.

## 3.Check Data & Clean Data

### 3.1. Check data column by column
- - -
**Check ID Column**

In [72]:
dataset1[(dataset1['Id'] < 9999999) | (dataset1['Id'] > 99999999)] #check ID range

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []

In [73]:
dataset1[dataset1.duplicated(['Id'])]

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []

According to the above checking, and the data description, 
>** Id: ** `is 8 digit Id of the job advertisement`, 

assert the ID is 8-digits number, and have no duplicate records.

**Check Title Column**

In [74]:
dataset1.Title.value_counts()

Mechanical Manufacturing Engineer  ****                                         1
Network Manager                                                                 1
Revenue Controller                                                              1
Sales Specialist  Telecomms  Nr Chepstow  ****k                                 1
ASSISTANT MANAGER  BRANDED PUB & RESTAURANT                                     1
Part Time Nursery Nurse Urgently Required                                       1
Head of Applications Delivery                                                   1
Commis Chef  Monday to Friday School Contract  14 Weeks Holiday P/A             1
Assistant Manager  Stunning High Volume Bar Venue  City, EC****  ****k          1
Property Sales Professionals / Senior Sales Negotiators                         1
Security Install / Service Engineer                                             1
Assistant Manager for a Stunning City Wine Bar                                  1
Compensation & A

In [75]:
print(dataset1.loc[19])
print(dataset1.loc[19]['Title'])

Id                                                           25452680
Title               Senior Sous Chef for **** rosette kitchen, up ...
Location                                           North East England
ContractType                                            not available
ContractTime                                            not available
Company                                               Clear Selection
Category                                  Hospitality & Catering Jobs
Salary per annum                                                26004
SourceName                                                caterer.com
OpenDate                                              20120101T120000
CloseDate                                             20120115T120000
Name: 19, dtype: object
Senior Sous Chef for **** rosette kitchen, up to ****


In [76]:
for record in dataset1.Title:
    if '*' in record:
        if 'k' in record or 'K' in record:
            print(record)

CHEF DE PARTIE POSITION IN **** ROSETTE HOTEL NYORKS  ****k
Senior Sous Chef for **** rosette kitchen, up to ****
General Manager  Funky, Cool Restaurant Concept  London  ****k 
Chef De Partie up to ****  Tips Ipswich Outskirts
Chef De Partie Norfolk Live In up to ****  Tips
Commis Chef  Milton Keynes **** package
Care Home Manager Job North London ****K
Staff Nurse (RGN) Doncaster ****K
Quality Manager – Nursing Homes Scotland ****K
RGN Blackpool **** per hour
Home Manager Inverness ****K
Home Manager Stockport ****K
Home Manager South Yorkshire ****K
Home Manager Barnsley ****K
General Manager Manchester ****k
Home Manager Clydebank ****K Bonus
Home Manager Aberdeen ****K BONUS
Care Home Manager Perth ****K
Home Manager Leeds ****K
Home Manager Rotherham ****K
Home Manager Fife ****K
Staff Nurse Stockton on Tees ****
RGN Deputy Bakewell **** p/h
Nurse Manager Altrincham ****K
Registered Nurse Blackpool ****
RGN Stockport Days, **** ph
Deputy Manager Middlesbrough ****K
RMN Nights Dar

Front End Web Developer JavaScript CSS HTML ****K BANKING
Outstanding Recruitment Opportunities OTE ****k  Nottingham
Industrial/Driving Hot Desk OTE ****k  Wakefield
Industrial/Driving Hot Desk OTE ****k  Warrington
Industrial/Driving Hot Desk OTE ****k  Wembley
2nd Line Desktop Support  **** users
Graduate Trainee Corporate Sales Executive ****K OTE
Sales & Marketing Executive  **** Commission  Property
Technical Sales and Service Representative  Manufacturing  ****K  Swindon
Business Development Manager /B2B /Surrey /BasicOTE ********k
QA Engineer  Madrid  €****K
POS/POP, Business Development Manager, ****k – ****k  Uncapped OTE  Car
Assistant Manager  Reading C****K
Deputy Manager: Outstanding quality brand C****K
Family Business Huge UK expansion seeks **** Working Partners pt / ft
Debt Management Consultant ****k OTE  Bolton
Field Sales Manager  ****k  Bonus  Benefits
Support Worker SOTH****
Support Worker EAST****
Support Workers (LB****)
Support Worker (LB****)
Pre Sales Engine

Net / CRM Developer, Surrey, ****K  Benefits
Dynamics AX Practice Lead, London, c****k
Project Worker  Mental Health x****
Customer Insight Analyst required c****k  top benefits
Sales Engineer  Chemicals up to ****k
Lead Business AnalystDebt FinanceLondon****K
Theatre Scrub Practitioner RGN London ****k p/a  London
Account Executive, Shopper Marketing, ****k
Dynamics CRM Business Analyst up to ****k  Benefits
Business Analyst  Regulatory Reporting  Tier **** Investment Bank
Junior Business Analyst  Tier **** Investment Bank  Market Risk
Internal IT Sales Executive  Milton Keynes ****k****k  OTE
IT Sales Consultant Milton Keynes ********
Internal Sales Engineer  Theale  ****k  IMMEDIATE OPENING
Telesales/ Telemarketers  Maidenhead up to ****k
Home Manager RGN Letchworth Garden City ****k  Hertford
Net / Java Software Developer Global Firm Manchester ****k
JavaScript Developer  Investment Bank  HTML ****
Anaesthetic Practitioner RGN London ****k p/a
Junior Telesales  London  ****K Basic 

It can be found that in the **Title** column, there are some special characters seem to have no meaning in this column, such as `up to ****`. By reading the title carefully, it can be found, some of this symbols should represent the salary, some of the symbols represet a number. it can be considered these characters may be replaced manually for privacy protection purpose. It should be retained in this case.

**Check the Location Column**

Next, we should check the `Location` column, based on the `describe()` method before, there are about 482 unique locations, There are several possible error for location data, such as: Typos, Location not exists(cannot be found), abbreviation etc.
- - -
We can first load the unique locations into a dataframe, with frequency of location.

In [77]:
locationCount = dataset1['Location'].value_counts().to_frame() #find unique location names

In [78]:
locationCount.head()

Location
UK                     3996
London                 2743
South East London      1433
The City                558
Leeds                   343

In [79]:
locationCount.columns = ['Count']

In [80]:
locationCount.head()

Count
UK                  3996
London              2743
South East London   1433
The City             558
Leeds                343

There are many locations in this dataframe, we should narrow the range. In practice, we can check the data with relatively low frequency, let's say, less than 10 times. (We assume data with high frequent relatively correct.) Now we create a new list contains data may need to be checked.

In [81]:
locCheck = locationCount[locationCount.Count<10].index.tolist()
locCheck[0:5]   #check the location appear less than 10 times

['Uxbridge', 'Frome', 'South Brent', 'Thame', 'Margate']

One way to find a location(address) in real life, it's using google map. In this task, we can use google map api instead. 
- - - 
**The advantages of this method**:
* find location precisely.
* fuzzy matching can find possible typos.
* large database support
- - -
**The disadvantages of this method**:
* can be time consuming.
* Free api have limitations.

In [82]:
gmaps = googlemaps.Client(key='AIzaSyC3GM3AP7cYIvsTCVcd0V5h7O06SzklhoA') #api key, may be replaced
geocode_result = gmaps.geocode('The City,UK') 
geocode_result     #get result, test api usage

[{'address_components': [{'long_name': 'City of London',
    'short_name': 'City of London',
    'types': ['locality', 'political', 'sublocality', 'sublocality_level_1']},
   {'long_name': 'London',
    'short_name': 'London',
    'types': ['locality', 'political']},
   {'long_name': 'Greater London',
    'short_name': 'Greater London',
    'types': ['administrative_area_level_2', 'political']},
   {'long_name': 'England',
    'short_name': 'England',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'United Kingdom',
    'short_name': 'GB',
    'types': ['country', 'political']}],
  'formatted_address': 'City of London, London, UK',
  'geometry': {'bounds': {'northeast': {'lat': 51.5233212, 'lng': -0.0727492},
    'southwest': {'lat': 51.5067659, 'lng': -0.113821}},
   'location': {'lat': 51.5123443, 'lng': -0.0909852},
   'location_type': 'APPROXIMATE',
   'viewport': {'northeast': {'lat': 51.5233212, 'lng': -0.0727492},
    'southwest': {'lat': 51.5067659,

We can use the `geocode` api, which can use a string to search location in google map, get the fully description of that location. From the above test, we can know how to get the geographic information.

Searching the string `The City,UK`, we can find location named `City of London`(using `formatted_address` key)
- - -
**We should also exclude some general locations we do not need to lookup, such as `London` `UK` `England` ** 

In [83]:
'''
This function may take sometime.
'''

for location in locCheck:
    if 'St.' not in location \
    and 'London' not in location\
    and 'England' not in location\
    and 'UK' not in location:      # exclude location don't need to look up
        geocode_result = gmaps.geocode(location+ ',UK')
        try:
            formatted_address = geocode_result[0]['formatted_address']
            if 'UK' in formatted_address and location.upper() not in formatted_address.upper(): #map the address
                print('Original Locations:', location) 
                print('May Correct Location:',formatted_address) #match possible right location name
        except:
            print(location)  #cannot find

In [84]:
dataset1[dataset1.Location == 'Docklands']

Id                                              Title   Location  \
2036   60517865  Graduate Customer Service Advisor Job  Docklan...  Docklands   
8065   67895479                            Payroll Sales Executive  Docklands   
8349   67944857  SOFTWARE TEST ENGINEER  Agile, TDD  Docklands,...  Docklands   
9956   68180003          Project Manager  Intranet  Banking  ****k  Docklands   
11803  68346992  Systems Administrator  (Windows)  London  ****...  Docklands   
13732  68544877  Engineer  Electrical  large data centre enviro...  Docklands   
20751  68996214           Global Vote Audit Analyst (Proxy Voting)  Docklands   
21865  69019190                Bookkeeper/Accounts with Quickbooks  Docklands   

        ContractType   ContractTime                       Company  \
2036       full_time      permanent                           NaN   
8065   not available      permanent  Carrington Blake Recruitment   
8349   not available      permanent                      Randstad   
9956   not available      permanent               Rullion IT Plus   
11803  not available      permanent                 Ashdown Group   
13732  not available      permanent                    Experis IT   
20751  not available      permanent                  Wells Tobias   
21865      part_time  not available                   Quay People   

                        Category Salary per annum            SourceName  \
2036   Accounting & Finance Jobs            21996        randstadfp.com   
8065                  Sales Jobs            21504      cv-library.co.uk   
8349            Engineering Jobs            42504  rengineeringjobs.com   
9956                     IT Jobs            62496         jobsite.co.uk   
11803                    IT Jobs            37500         jobsite.co.uk   
13732                    IT Jobs            45000         jobsite.co.uk   
20751  Accounting & Finance Jobs            21000      cv-library.co.uk   
21865                    IT Jobs             9996      cv-library.co.uk   

              OpenDate        CloseDate  
2036   20130705T150000  20130804T150000  
8065   20120404T150000  20120703T150000  
8349   20130915T120000  20131015T120000  
9956   20130508T120000  20130806T120000  
11803  20130620T150000  20130819T150000  
13732  20120326T120000  20120525T120000  
20751  20130202T150000  20130503T150000  
21865  20130404T120000  20130603T120000

In [85]:
dataset1[dataset1.Location == 'Surrey'][0:1]

Id                                              Title Location  \
3  12613049  Engineering Systems Analyst / Mathematical Mod...   Surrey   

    ContractType ContractTime                       Company          Category  \
3  not available    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  
3            27504  cv-library.co.uk  20121214T000000  20130314T000000

In [86]:
dataset1[dataset1.Location == 'Surey'][0:1]

Id                                Title Location ContractType  \
122  46626802  Nursing Home Staff Nurse Job Surrey    Surey    full_time   

      ContractTime Company                   Category Salary per annum  \
122  not available     NaN  Healthcare & Nursing Jobs            23040   

         SourceName         OpenDate        CloseDate  
122  careworx.co.uk  20130512T120000  20130810T120000

It can actually find some typos in the location. Such as: `Surey` -> `Surrey`,`Nottinham` -> `Nottingham`.

However, the `Docklands` should not be seemed as a error, the reason is that: This Location can be found in the website, In UK(London), this location may not be found in google maps as an accurate address. refer to [London Docklands](https://en.wikipedia.org/wiki/London_Docklands)

**Check Contract type & time**

In [87]:
dataset1.ContractType.value_counts()

not available    19499
full_time         4883
part_time          695
Name: ContractType, dtype: int64

In [88]:
dataset1.ContractTime.value_counts()

permanent        16194
not available     6212
contract          2671
Name: ContractTime, dtype: int64

From the above checking, it seems there are no obvious error, however, according to the specification, we can slightly change the `not availabe` to the term `non_specified`, we can consider it later.

**Check Company column**

In [89]:
dataset1[dataset1.Company.isnull()]

Id                                              Title  \
51     34479775                                  Credit Controller   
55     34790490                                     Java Developer   
66     38209887                     Transaction Services Executive   
67     38209888                                    Audit Executive   
68     38209893                            Audit Assistant Manager   
70     38758190                                  Account Executive   
78     40880208                                 Compliance Manager   
81     41879934   Future Leaders Development Programme  Technology   
105    46626508                        Support Worker Job Chepstow   
106    46626509                     Support Worker Job Abergavenny   
107    46626511                          Staff Nurse Job Berkshire   
108    46626514                    Pharmacy Technician Job Norwich   
109    46626520                         Support Worker Job Cardiff   
110    46626537               RGN Registered Nurse Job Southampton   
111    46626546                      RGN Nurse Job Bury St Edmunds   
112    46626547                          Scrub Nurse Job Braintree   
113    46626549                Healthcare Assistant Job Nottingham   
114    46626555       Consultant Clinical Psychologist Job Norwich   
115    46626557  Operating Department Practitioner ODP Jobs Not...   
116    46626559                 RGN Prison Nurse Job Staffordshire   
117    46626566             Clinical Research Nurse Job Nottingham   
118    46626571                    RGN Registered Nurse Job Durham   
119    46626573     Prison Nurse Practitioner Jobs Nottinghamshire   
120    46626574    Care Assistants and Support Workers Job Glasgow   
121    46626799           Saturday only Night Nurse Job Winchester   
122    46626802                Nursing Home Staff Nurse Job Surrey   
123    46626826                        Substance Misuse Nurse Kent   
124    46626831        Registered Mental Health Nurse (RMN) Durham   
125    46626832    Registered Mental Health Nurse (RMN) Sunderland   
126    46626833          Registered General Nurse (RGN) Sunderland   
...         ...                                                ...   
23862  69166844                       Nurse : Surgical Staff Nurse   
23863  69166851           Nurse : Deputy Nursing Home Manager ****   
23864  69167060  Nurse : Registered General Nurse Jobs in Leigh...   
24582  69189730            Corporate Tax Compliance Senior Manager   
24586  69190364                        Inclusion Support Assistant   
24592  69190507      Junior Database Administrator  Holton St Mary   
24596  69190934                                Graphics Technician   
24629  69191688                  Accounts Assistant (MultiLingual)   
24631  69191782             Domestic Violence/Abuse Support Worker   
24633  69192290                        Progress Webspeed Developer   
24634  69192326  Property Sales Consultant / Senior Sales Negot...   
24635  69192551            FT Skilled Senior Carers  Days & Nights   
24653  69193530                                   **** x Caretaker   
24686  69196406  APPOINTMENT OF INDEPENDENT MEMBERS TO CARDIFF ...   
24723  69200075                         Trainee Marketing Advisors   
24752  69203584                                  Dividends Analyst   
24753  69203586                                Settlements Analyst   
24754  69203620                               Client Admin Support   
24825  69208639         Resourcing Coordinator (Full or Part Time)   
24921  69215963                  Solar Sales Consultant  Inverness   
24923  69216134                            Skilled Specialist Chef   
24924  69216174                 Project Manager (Supported Living)   
25008  69220336                           Trainer  NVQ Level 3 BIT   
25009  69220384           Appointment Maker  Temporary / Permanent   
25012  69220596               Promotions Sales Manager  Short Term   
25037  69227935  Commun

It can be found in the above checking result, It is nearly impossible to fill the missing data for the company, the reason is that we cannot get any information from other columns for the Company's data. We can keep the missing data, refer as not filled or not available.

**Check Category**

In [90]:
dataset1.Category.value_counts()

IT Jobs                             7085
Healthcare & Nursing Jobs           4334
Engineering Jobs                    3458
Accounting & Finance Jobs           3099
Sales Jobs                          2609
Hospitality & Catering Jobs         2124
Teaching Jobs                       1378
PR, Advertising & Marketing Jobs     990
Name: Category, dtype: int64

From the above checking, it seems there are no obvious error for this column.

**Check Salary**

In [91]:
dataset1[dataset1['Salary per annum'].str.contains('\D')] # salary contain non-numeric characters 

Id                                              Title  \
133    46626860            Registered Home Manager Job Bournemouth   
238    46627928           Care Home Manager Job North London ****K   
305    46628805                     Home Care Workers Berkhampsted   
596    46634306                  Home Manager Mental Health 11 bed   
647    46634923        RGN Nurse Hull Days or Nights **** per hour   
830    46637596                     Staff Nurse South Shields ****   
896    48082563  Senior Chef de Partie  One AA Rosette Hotel  T...   
952    49065458              Service Manager Learning Disabilities   
980    49689021                                 Assessment Officer   
996    49845058  Chef de Partie  Fresh Food Pub  Good Reputatio...   
1062   51061201              Community Home Workers Melton Mowbray   
1081   51285942                     Senior Social Worker Fostering   
1141   52489008                                      RGN Care Home   
1292   55400384                    Live in Care Worker Jobs Exeter   
1379   55408955                      Senior Applications Developer   
1410   55409587                          MVNO Enterprise Architect   
1525   56209766                                 Nurse Advisor COPD   
1621   56814478                   Chef de Partie  **** AA Rosettes   
1797   58533485          Senior Embedded Controls Systems Engineer   
1822   58759123  Helpdesk / Desktop Support analystExcel, Word,...   
2051   60684270              Door to door / Field Sales Canvassers   
2063   60685515          Staff Nurse Elderly Care Stockton on Tees   
2118   61445890                              Mental Health Advisor   
2602   64145428                             DEPUTY MANAGER READING   
2724   64573377                 Deputy Care Manager / Care Manager   
2889   64809245  Contract Mechanical Draughtsperson (Huddersfield)   
2923   64813318  Manufacturing Engineer (Mechanical / Pumps / CNC)   
3057   65171402                                 Complaints Manager   
3186   65188754       Technical Sales Engineer (ElectroMechanical)   
3274   65317743  Assistant General Manager  New Opening  Aberde...   
...         ...                                                ...   
21666  69012121                    Adult Speech Language Therapist   
21734  69017589  BI Manager, SQL Server Financial ServicesHertf...   
21744  69018023  PHP WEB DEVELOPER / PHP PROGRAMMER / SOFTWARE ...   
21951  69022222  Software Test Engineer **** East Anglia  Embedded   
22029  69022816         PR and Events Manager (Maternity Contract)   
22283  69036978  Application Support Analyst (legal sector)  Lo...   
22418  69039080          Oracle & SQL DBA  Investment Bank  London   
22420  69039153  Business Analyst (Debt Finance, Corporate Bank...   
22465  69039760  Project Manager (Software development) near Ma...   
22493  69039997        Graduate Software Developer – online gaming   
23111  69080469                                  Catering Engineer   
23152  69080768              Drilling Equipment Service Technician   
23207  69081257                    Principal/Senior ITS Consultant   
23365  69083476                                  Auto CAD Designer   
23478  69091714                                Web Support Analyst   
23557  69118140  Sales Manager Drain Cleaning & Inspection Equi...   
23619  69122549                                   Credit Assistant   
23815  69162548                             Quality/ CMM Inspector   
23984  69169854                   Marketing Manager  Food industry   
24008  69170235                         Secretarial/PA Team Leader   
24027  69170445  Graduate Trainee Financial Risk Trader  ****k ...   
24113  69171949         DEMI CHEF DE PARTIE  ****  excellent tips    
24170  69172786         Qualified Social Workers for ALL of London   
24319  69182099                                       Bars Manager   
24352  69182395  General Manager Leisure Complex SE London Area...   
24364  69182561        

In [92]:
len(dataset1[dataset1['Salary per annum'].str.contains('\D')])

200

From the above checking, we can find that there are some data not numeric type, refer to the specification: 
> ** `Salary per annum` is annual salary of the advertised job position, e.g., 80000 **
- - - 
We should clean the data to make data structure normalized. For instance: `30K` to `30000`,`30871.199999999997 - 34120.8` to `30871.2 - 34120.8`. The reason for this change is: salary should be numeric type in this data, we also should keep the salary range formmat cause it is reasonable in real life, the long float format is unnecessary for salary. 

**Check SourceName**
- - -
> The website where the job position is advertised.

In [93]:
'''
check source Name
'''
for record in dataset1.SourceName.unique().tolist():
    if re.search('.*.co',record) is None \
    and re.search('.*.net',record) is None\
    and re.search('.*.org',record) is None:
        print(record)

monashstudent
jobcareer


In [94]:
'''
check source Name
'''
for record in dataset1.SourceName.unique().tolist():
    if re.search('@',record) is not None:
        print(record)   

admin@caterer.com


In [95]:
dataset1[dataset1.SourceName == 'admin@caterer.com']

Id                                             Title  \
15379  68672352  Digital Account Manager  **** plus a Great Bonus   

                Location   ContractType ContractTime      Company    Category  \
15379  South East London  not available    permanent  Blu Digital  Sales Jobs   

      Salary per annum         SourceName         OpenDate        CloseDate  
15379            27504  admin@caterer.com  20120516T000000  20120715T000000

In [96]:
dataset1[dataset1.SourceName == 'jobcareer']

Id                                            Title Location  \
12871  68393935  Killer Javascript role Exclusive to Brightwater  Belfast   

      ContractType ContractTime            Company Category Salary per annum  \
12871    full_time    permanent  Brightwater Group  IT Jobs            39996   

      SourceName         OpenDate        CloseDate  
12871  jobcareer  20120205T150000  20120306T150000

In [97]:
dataset1[dataset1.SourceName == 'monashstudent']

Id                                          Title Location  \
5663  66932999  Registered Midwives RM  Lincolnshire  Lincoln  Lincoln   

     ContractType   ContractTime        Company                   Category  \
5663    part_time  not available  The A24 Group  Healthcare & Nursing Jobs   

     Salary per annum     SourceName         OpenDate        CloseDate  
5663            40140  monashstudent  20130124T150000  20130223T150000

How should we define a valid website, we can use some packages try to open the url indeed, however, this data comes from about 5 years ago, there may be some websites cannot be opened but existed. We can try to use regular expression to match the websites that may be a valid address. For example, a website should contain key words like `.com`, `.net`, `.org` etc. There are actually two records not valid website.

There is another wired record, which use admin@caterer.com, which seem to be a e-mail address,it may be replaced by the website itself instead.

**Check Date**
- - -
Refer to the specification:
> The opening time for applying for the advertised job position, e.g., 20120104T150000, means 3pm, 4th January 2012.
- - -
> The closing time for applying for the advertised job position, e.g., 20120104T150000, means 3pm, 4th January 2012.

We can try to prase the date with the format, if we have some error, we can find out using `try except block`

In [98]:
'''
check Open Date
'''
for date in dataset1.OpenDate:
    try:
        datetime.datetime.strptime(date,"%Y%m%dT%H%M%S")  #try to prase date
    except:
        print(date) #when exception occur, print wrong date

20131803T120000
20132606T000000
20122003T150000
20121512T150000
20133004T150000
20131908T000000
20121406T120000
20131509T000000
20132901T150000
20132108T120000


In [99]:
'''
check Close Date
'''
for date in dataset1.CloseDate:
    try:
        datetime.datetime.strptime(date,"%Y%m%dT%H%M%S")
    except:
        print(date)

It can be found that there are some error data in the Open Data column. Choosing one of the record as a sample:

`20131803T120000` the month in this record is `18`, which is impossible in real life. There is a high chance that the error data appear because of the wrong order of the month and day. `20131803T120000` should be `20130318T120000`

Another potential error for date according to the meaning is that close date should always later than opendate. It can be checked in the cleaning step.

### 3.2. Try to clean data column by column

**Cleaning Location Column**
- - -
Change the typos which we found above.

In [100]:
replacedict = {'Leads' : 'Leeds','Reeding' : 'Reading','Surey' : 'Surrey','Oxfords' : 'Oxford','Nottinham' : 'Nottingham'}
dataset1['Location'].replace(replacedict,inplace = True) #replace location names

In [101]:
print(dataset1[dataset1.Location == 'Leads'])
print(dataset1[dataset1.Location == 'Reeding'])
print(dataset1[dataset1.Location == 'Surey'])
print(dataset1[dataset1.Location == 'Oxfords'])
print(dataset1[dataset1.Location == 'Nottinham'])  #check results

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []
Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []
Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []
Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []
Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate]
Index: []


**Cleaning Contract Columns**

In [102]:
# not available == non_specified
dataset1['ContractType'].replace({'not available':'non_specified'},inplace = True)
dataset1['ContractTime'].replace({'not available':'non_specified'},inplace = True)

**Cleaning Salary Column**
 - - -
 We can replace `K` with `000`
 
 Using a function to unify the float numbers

In [103]:
dataset1[dataset1['Salary per annum'].str.contains('\D')]

Id                                              Title  \
133    46626860            Registered Home Manager Job Bournemouth   
238    46627928           Care Home Manager Job North London ****K   
305    46628805                     Home Care Workers Berkhampsted   
596    46634306                  Home Manager Mental Health 11 bed   
647    46634923        RGN Nurse Hull Days or Nights **** per hour   
830    46637596                     Staff Nurse South Shields ****   
896    48082563  Senior Chef de Partie  One AA Rosette Hotel  T...   
952    49065458              Service Manager Learning Disabilities   
980    49689021                                 Assessment Officer   
996    49845058  Chef de Partie  Fresh Food Pub  Good Reputatio...   
1062   51061201              Community Home Workers Melton Mowbray   
1081   51285942                     Senior Social Worker Fostering   
1141   52489008                                      RGN Care Home   
1292   55400384                    Live in Care Worker Jobs Exeter   
1379   55408955                      Senior Applications Developer   
1410   55409587                          MVNO Enterprise Architect   
1525   56209766                                 Nurse Advisor COPD   
1621   56814478                   Chef de Partie  **** AA Rosettes   
1797   58533485          Senior Embedded Controls Systems Engineer   
1822   58759123  Helpdesk / Desktop Support analystExcel, Word,...   
2051   60684270              Door to door / Field Sales Canvassers   
2063   60685515          Staff Nurse Elderly Care Stockton on Tees   
2118   61445890                              Mental Health Advisor   
2602   64145428                             DEPUTY MANAGER READING   
2724   64573377                 Deputy Care Manager / Care Manager   
2889   64809245  Contract Mechanical Draughtsperson (Huddersfield)   
2923   64813318  Manufacturing Engineer (Mechanical / Pumps / CNC)   
3057   65171402                                 Complaints Manager   
3186   65188754       Technical Sales Engineer (ElectroMechanical)   
3274   65317743  Assistant General Manager  New Opening  Aberde...   
...         ...                                                ...   
21666  69012121                    Adult Speech Language Therapist   
21734  69017589  BI Manager, SQL Server Financial ServicesHertf...   
21744  69018023  PHP WEB DEVELOPER / PHP PROGRAMMER / SOFTWARE ...   
21951  69022222  Software Test Engineer **** East Anglia  Embedded   
22029  69022816         PR and Events Manager (Maternity Contract)   
22283  69036978  Application Support Analyst (legal sector)  Lo...   
22418  69039080          Oracle & SQL DBA  Investment Bank  London   
22420  69039153  Business Analyst (Debt Finance, Corporate Bank...   
22465  69039760  Project Manager (Software development) near Ma...   
22493  69039997        Graduate Software Developer – online gaming   
23111  69080469                                  Catering Engineer   
23152  69080768              Drilling Equipment Service Technician   
23207  69081257                    Principal/Senior ITS Consultant   
23365  69083476                                  Auto CAD Designer   
23478  69091714                                Web Support Analyst   
23557  69118140  Sales Manager Drain Cleaning & Inspection Equi...   
23619  69122549                                   Credit Assistant   
23815  69162548                             Quality/ CMM Inspector   
23984  69169854                   Marketing Manager  Food industry   
24008  69170235                         Secretarial/PA Team Leader   
24027  69170445  Graduate Trainee Financial Risk Trader  ****k ...   
24113  69171949         DEMI CHEF DE PARTIE  ****  excellent tips    
24170  69172786         Qualified Social Workers for ALL of London   
24319  69182099                                       Bars Manager   
24352  69182395  General Manager Leisure Complex SE London Area...   
24364  69182561        

In [104]:
dataset1['Salary per annum'].replace('K','000',regex = True,inplace = True)

In [105]:
def unifyFloat(num):
    nums = num.split('-')
    first = round(float(nums[0]), 2)
    second = round(float(nums[1]), 2)
    return str(first) + ' - ' + str(second)  #unify float numbers

In [106]:
dataset1.loc[dataset1['Salary per annum'].str.contains('\D'),'Salary per annum'] = \
dataset1[dataset1['Salary per annum'].str.contains('\D')]['Salary per annum'].apply(lambda x: unifyFloat(x))

**Check the results**

In [107]:
dataset1[dataset1['Salary per annum'].str.contains('\D')]

Id                                              Title  \
647    46634923        RGN Nurse Hull Days or Nights **** per hour   
896    48082563  Senior Chef de Partie  One AA Rosette Hotel  T...   
980    49689021                                 Assessment Officer   
1062   51061201              Community Home Workers Melton Mowbray   
1141   52489008                                      RGN Care Home   
1292   55400384                    Live in Care Worker Jobs Exeter   
1525   56209766                                 Nurse Advisor COPD   
1797   58533485          Senior Embedded Controls Systems Engineer   
1822   58759123  Helpdesk / Desktop Support analystExcel, Word,...   
2051   60684270              Door to door / Field Sales Canvassers   
2118   61445890                              Mental Health Advisor   
2602   64145428                             DEPUTY MANAGER READING   
2724   64573377                 Deputy Care Manager / Care Manager   
2889   64809245  Contract Mechanical Draughtsperson (Huddersfield)   
2923   64813318  Manufacturing Engineer (Mechanical / Pumps / CNC)   
3057   65171402                                 Complaints Manager   
3186   65188754       Technical Sales Engineer (ElectroMechanical)   
3587   65879969                               Contract Salesperson   
3927   66197678  Front End Web Developer – (HTML, CSS, Javascript)   
3973   66367073                      Financial Controller (Assets)   
4705   66608673  INFRASTRUCTURE SUPPORT – OPERATIONS – CLOUD, S...   
5708   66935869         Account Manager, Account Executive  Energy   
5988   67017625                      Contracts Manager (Permanent)   
6225   67153390   Electrical Field Service Engineer x**** (Cranes)   
6270   67208324            Building Services Supervisor in Chelsea   
6710   67333923                               HVAC Design Engineer   
6977   67389276                     Lead Business Analyst, Banking   
7141   67638912  IT Sales Executive / New Business Development ...   
7296   67652461  MS SharePoint Senior Developer, R****k, Jo bur...   
7347   67652829               IT Equipment Installations Assistant   
...         ...                                                ...   
17906  68702082  Electronic / Electrical Test Engineer (Lithium...   
17983  68703232                             Commercial Team Leader   
18002  68703556             Service Desk / Desktop Support Analyst   
18174  68706043  Manager/ Unit Manager/ Store Manager/ Catering...   
18905  68715405  TECHNICAL LEAD CASPNET MVC SCRUM TDD HOSPITALI...   
18964  68716258       Project Manager  Southampton  ****K to ****K   
19270  68787221  IT Sales Executive  Accrington  ****k to ****k...   
19445  68804934                   Mechanical Estimator  Birmingham   
19548  68805804          Relocate to the US? QA Software Engineers   
20070  68834330                Field Service Connectivity Engineer   
20175  68837150                Restaurant Manager  ****AA Rosettes   
20854  68999922          INTERNAL SALES  **** Year Maternity Cover   
21064  69004198                           Helpdesk Apprenticeships   
21242  69005600            Outsource Manager / Employer Engagement   
21247  69005706  FRENCH AND SPANISH TEACHER  QTS  **** **** PER...   
21313  69006748                           Part Time Year 6 Teacher   
21734  69017589  BI Manager, SQL Server Financial ServicesHertf...   
22283  69036978  Application Support Analyst (legal sector)  Lo...   
22418  69039080          Oracle & SQL DBA  Investment Bank  London   
23111  69080469                                  Catering Engineer   
23152  69080768              Drilling Equipment Service Technician   
23207  69081257                    Principal/Senior ITS Consultant   
23365  69083476                                  Auto CAD Designer   
23557  69118140  Sales Manager Drain Cleaning & Inspection Equi...   
23619  69122549                                   Credit Assistant   
23815  69162548        

**Cleaning the SourceName column**
- - - 
We can replace `monashstudent` with `monash.edu/students`

replace `jobcareer` with `jobcareer.com`

replace "admin@caterer.com" with `caterer.com`

In [108]:
replacesource = {'monashstudent' : 'monash.edu/students','jobcareer' : 'jobcareer.com','admin@caterer.com' : 'caterer.com'}
dataset1['SourceName'].replace(replacesource,inplace = True) #replace source name

**Cleaning the Date columns**
- - -
We can try to make another two columns use to store the paresed date, if we have error, we can use the variable `errors="coerce"` to keep the error as null value.

In [109]:
dataset1['OpenDate_parsed'] = pd.to_datetime(dataset1.OpenDate,format="%Y%m%dT%H%M%S",errors='coerce') #leave error date as Null value

In [110]:
dataset1[dataset1['OpenDate_parsed'].isnull()]

Id                                              Title  \
1102   51593546                       Community Nurse South London   
2104   61282006                           Deputy Home Care Manager   
2839   64796836  Technical Tester  Manual, Automation, Java, SQ...   
5707   66935867  Technical Sales Manager, Pigments Fillers to S...   
10881  68257270   Desire a Change?  London  ****K Basic  ****K OTE   
11948  68350622                  Mechanical Design Engineer (HVAC)   
15353  68671948                                 Investment Analyst   
22918  69072305                                     Scheme Actuary   
23007  69079402                Naval architect/structural engineer   
23169  69080910  MS (Elec biased) Maintenance Engineer / Superv...   

                  Location   ContractType   ContractTime  \
1102              The City  non_specified  non_specified   
2104                Barnet  non_specified  non_specified   
2839                London  non_specified      permanent   
5707                    UK  non_specified  non_specified   
10881    South East London  non_specified  non_specified   
11948         Gainsborough  non_specified      permanent   
15353             Cheshire  non_specified      permanent   
22918            Hampshire  non_specified      permanent   
23007  Newcastle Upon Tyne  non_specified      permanent   
23169            Harrogate  non_specified      permanent   

                      Company                   Category Salary per annum  \
1102                      NaN  Healthcare & Nursing Jobs            29004   
2104                      NaN  Healthcare & Nursing Jobs            20004   
2839                      NaN                    IT Jobs            19500   
5707   Wallace Hind Selection                 Sales Jobs            45000   
10881      Zest 2 Recruitment                 Sales Jobs            13500   
11948                     NaN                    IT Jobs            39996   
15353         Idex Consulting  Accounting & Finance Jobs            24000   
22918        Sammons Pensions  Accounting & Finance Jobs            54996   
23007    Workshop Recruitment           Engineering Jobs            33504   
23169     Jenrick Engineering           Engineering Jobs            17004   

                         SourceName         OpenDate        CloseDate  \
1102                 careworx.co.uk  20131803T120000  20130417T120000   
2104                 careworx.co.uk  20132606T000000  20130710T000000   
2839            theitjobboard.co.uk  20122003T150000  20120519T150000   
5707              salestarget.co.uk  20121512T150000  20130213T150000   
10881             salestarget.co.uk  20133004T150000  20130530T150000   
11948              technojobs.co.uk  20131908T000000  20131117T000000   
15353                 totaljobs.com  20121406T120000  20120912T120000   
22918  professionalpensionsjobs.com  20131509T000000  20131114T000000   
23007              cv-library.co.uk  20132901T150000  20130429T150000   
23169              cv-library.co.uk  20132108T120000  20131119T120000   

      OpenDate_parsed  
1102              NaT  
2104              NaT  
2839              NaT  
5707              NaT  
10881             NaT  
11948             NaT  
15353             NaT  
22918             NaT  
23007             NaT  
23169             NaT

**Define a replace function and test if it will work**

In [111]:
def fixDate(date):
    pattern = re.findall('(\d{4})(\d{2})(\d{2})(T\d{2}\d{2}\d{2})',date)
    return pattern[0][0] + pattern[0][2] + pattern[0][1] + pattern[0][3]

In [112]:
fixDate('20131803T120000')

'20130318T120000'

Get the error data index list

In [113]:
indexList = dataset1[dataset1['OpenDate_parsed'].isnull()].index

Fix the errors

In [114]:
for i in indexList:
    dataset1.at[i,'OpenDate'] = fixDate(dataset1.loc[i]['OpenDate'])

In [115]:
dataset1['OpenDate_parsed'] = pd.to_datetime(dataset1.OpenDate,format="%Y%m%dT%H%M%S",errors='coerce') 
dataset1[dataset1['OpenDate_parsed'].isnull()]

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate, OpenDate_parsed]
Index: []

Check the close date Format 

In [116]:
dataset1['CloseDate_parsed'] = pd.to_datetime(dataset1.CloseDate,format="%Y%m%dT%H%M%S",errors='coerce')

In [117]:
dataset1[dataset1['CloseDate_parsed'].isnull()]

Empty DataFrame
Columns: [Id, Title, Location, ContractType, ContractTime, Company, Category, Salary per annum, SourceName, OpenDate, CloseDate, OpenDate_parsed, CloseDate_parsed]
Index: []

In [118]:
dataset1.head()

Id                                              Title   Location  \
0  12612628                        Engineering Systems Analyst    Dorking   
1  12612830                            Stress Engineer Glasgow    Glasgow   
2  12612844                   Modelling and simulation analyst  Hampshire   
3  12613049  Engineering Systems Analyst / Mathematical Mod...     Surrey   
4  12613647         Pioneer, Miser Engineering Systems Analyst     Surrey   

    ContractType ContractTime                       Company          Category  \
0  non_specified    permanent  Gregory Martin International  Engineering Jobs   
1  non_specified    permanent  Gregory Martin International  Engineering Jobs   
2  non_specified    permanent  Gregory Martin International  Engineering Jobs   
3  non_specified    permanent  Gregory Martin International  Engineering Jobs   
4  non_specified    permanent  Gregory Martin International  Engineering Jobs   

  Salary per annum        SourceName         OpenDate        CloseDate  \
0            24996  cv-library.co.uk  20121103T000000  20121203T000000   
1            30000  cv-library.co.uk  20130108T150000  20130408T150000   
2            30000  cv-library.co.uk  20130726T150000  20130924T150000   
3            27504  cv-library.co.uk  20121214T000000  20130314T000000   
4            24996  cv-library.co.uk  20131025T000000  20131224T000000   

      OpenDate_parsed    CloseDate_parsed  
0 2012-11-03 00:00:00 2012-12-03 00:00:00  
1 2013-01-08 15:00:00 2013-04-08 15:00:00  
2 2013-07-26 15:00:00 2013-09-24 15:00:00  
3 2012-12-14 00:00:00 2013-03-14 00:00:00  
4 2013-10-25 00:00:00 2013-12-24 00:00:00

Find the error which close date earlier than open date

In [119]:
dataset1[dataset1.OpenDate_parsed > dataset1.CloseDate_parsed] #unavailable date

Id                                              Title  \
6659   67330693                    Pensions Implementation Manager   
9568   68091856                       Manager Technical Operations   
11043  68290941                       Planning Enforcement officer   
12473  68360715              Penetration Tester / Reverse Engineer   
19142  68729089              Registered Home Manager  Loughborough   
24206  69173205  User Interface/UI Developer  JavaScript, jQuer...   
24297  69181738  Registered General Nurse (RGN)/ Registered Men...   
25039  69228213  SCIENTIFIC (PHYSICS, OR CHEMISTRY, OR ELECTRON...   

             Location   ContractType   ContractTime                   Company  \
6659         Brighton  non_specified      permanent                       NaN   
9568   Shepton Mallet  non_specified      permanent         Index Recruitment   
11043          London      full_time       contract                  Randstad   
12473              UK  non_specified       contract          Vertex Solutions   
19142    Loughborough      full_time  non_specified    Compass Associates Ltd   
24206          London  non_specified      permanent            Dawson & Walsh   
24297           Poole      full_time  non_specified  Executive Care Group Ltd   
25039      Warrington      full_time      permanent                       NaN   

                        Category Salary per annum  \
6659   Accounting & Finance Jobs            35004   
9568                     IT Jobs            62004   
11043           Engineering Jobs            39360   
12473                    IT Jobs            76908   
19142  Healthcare & Nursing Jobs            42504   
24206                    IT Jobs            45000   
24297  Healthcare & Nursing Jobs            24960   
25039           Engineering Jobs            24996   

                        SourceName         OpenDate        CloseDate  \
6659                    hays.co.uk  20130703T000000  20130404T000000   
9568                 totaljobs.com  20120306T000000  20120205T000000   
11043  jobs.planningresource.co.uk  20131105T000000  20131006T000000   
12473                 cwjobs.co.uk  20131222T150000  20131023T150000   
19142               staffnurse.com  20120803T150000  20120505T150000   
24206                totaljobs.com  20130622T150000  20130324T150000   
24297               staffnurse.com  20130116T000000  20121018T000000   
25039                  fish4.co.uk  20130708T120000  20130608T120000   

          OpenDate_parsed    CloseDate_parsed  
6659  2013-07-03 00:00:00 2013-04-04 00:00:00  
9568  2012-03-06 00:00:00 2012-02-05 00:00:00  
11043 2013-11-05 00:00:00 2013-10-06 00:00:00  
12473 2013-12-22 15:00:00 2013-10-23 15:00:00  
19142 2012-08-03 15:00:00 2012-05-05 15:00:00  
24206 2013-06-22 15:00:00 2013-03-24 15:00:00  
24297 2013-01-16 00:00:00 2012-10-18 00:00:00  
25039 2013-07-08 12:00:00 2013-06-08 12:00:00

To find out how to fix the error, we can create another column to store duration for each records.

In [120]:
dataset1['duration'] = (dataset1.CloseDate_parsed - dataset1.OpenDate_parsed)

In [121]:
dataset1.duration.describe()

count                      25077
mean     53 days 04:26:50.718985
std      25 days 19:15:46.166444
min           -90 days +00:00:00
25%             30 days 00:00:00
50%             60 days 00:00:00
75%             60 days 00:00:00
max             90 days 00:00:00
Name: duration, dtype: object

In [122]:
dataset1[dataset1.OpenDate_parsed > dataset1.CloseDate_parsed]

Id                                              Title  \
6659   67330693                    Pensions Implementation Manager   
9568   68091856                       Manager Technical Operations   
11043  68290941                       Planning Enforcement officer   
12473  68360715              Penetration Tester / Reverse Engineer   
19142  68729089              Registered Home Manager  Loughborough   
24206  69173205  User Interface/UI Developer  JavaScript, jQuer...   
24297  69181738  Registered General Nurse (RGN)/ Registered Men...   
25039  69228213  SCIENTIFIC (PHYSICS, OR CHEMISTRY, OR ELECTRON...   

             Location   ContractType   ContractTime                   Company  \
6659         Brighton  non_specified      permanent                       NaN   
9568   Shepton Mallet  non_specified      permanent         Index Recruitment   
11043          London      full_time       contract                  Randstad   
12473              UK  non_specified       contract          Vertex Solutions   
19142    Loughborough      full_time  non_specified    Compass Associates Ltd   
24206          London  non_specified      permanent            Dawson & Walsh   
24297           Poole      full_time  non_specified  Executive Care Group Ltd   
25039      Warrington      full_time      permanent                       NaN   

                        Category Salary per annum  \
6659   Accounting & Finance Jobs            35004   
9568                     IT Jobs            62004   
11043           Engineering Jobs            39360   
12473                    IT Jobs            76908   
19142  Healthcare & Nursing Jobs            42504   
24206                    IT Jobs            45000   
24297  Healthcare & Nursing Jobs            24960   
25039           Engineering Jobs            24996   

                        SourceName         OpenDate        CloseDate  \
6659                    hays.co.uk  20130703T000000  20130404T000000   
9568                 totaljobs.com  20120306T000000  20120205T000000   
11043  jobs.planningresource.co.uk  20131105T000000  20131006T000000   
12473                 cwjobs.co.uk  20131222T150000  20131023T150000   
19142               staffnurse.com  20120803T150000  20120505T150000   
24206                totaljobs.com  20130622T150000  20130324T150000   
24297               staffnurse.com  20130116T000000  20121018T000000   
25039                  fish4.co.uk  20130708T120000  20130608T120000   

          OpenDate_parsed    CloseDate_parsed duration  
6659  2013-07-03 00:00:00 2013-04-04 00:00:00 -90 days  
9568  2012-03-06 00:00:00 2012-02-05 00:00:00 -30 days  
11043 2013-11-05 00:00:00 2013-10-06 00:00:00 -30 days  
12473 2013-12-22 15:00:00 2013-10-23 15:00:00 -60 days  
19142 2012-08-03 15:00:00 2012-05-05 15:00:00 -90 days  
24206 2013-06-22 15:00:00 2013-03-24 15:00:00 -90 days  
24297 2013-01-16 00:00:00 2012-10-18 00:00:00 -90 days  
25039 2013-07-08 12:00:00 2013-06-08 12:00:00 -30 days

Consider the above statistics, we can find that the duration for the errors seem normal except they are negative. These errors may come from the exchange of open-date and close-date, we can define a method to swap these two date.

In [123]:
indexList1 = dataset1[dataset1.OpenDate_parsed > dataset1.CloseDate_parsed].index

In [124]:
for i in indexList1:
    openDate = dataset1.loc[i]['OpenDate']
    closeDate = dataset1.loc[i]['CloseDate']
    dataset1.loc[i,'OpenDate'] = closeDate
    dataset1.loc[i,'CloseDate'] = openDate

In [125]:
dataset1[dataset1.OpenDate_parsed > dataset1.CloseDate_parsed]

Id                                              Title  \
6659   67330693                    Pensions Implementation Manager   
9568   68091856                       Manager Technical Operations   
11043  68290941                       Planning Enforcement officer   
12473  68360715              Penetration Tester / Reverse Engineer   
19142  68729089              Registered Home Manager  Loughborough   
24206  69173205  User Interface/UI Developer  JavaScript, jQuer...   
24297  69181738  Registered General Nurse (RGN)/ Registered Men...   
25039  69228213  SCIENTIFIC (PHYSICS, OR CHEMISTRY, OR ELECTRON...   

             Location   ContractType   ContractTime                   Company  \
6659         Brighton  non_specified      permanent                       NaN   
9568   Shepton Mallet  non_specified      permanent         Index Recruitment   
11043          London      full_time       contract                  Randstad   
12473              UK  non_specified       contract          Vertex Solutions   
19142    Loughborough      full_time  non_specified    Compass Associates Ltd   
24206          London  non_specified      permanent            Dawson & Walsh   
24297           Poole      full_time  non_specified  Executive Care Group Ltd   
25039      Warrington      full_time      permanent                       NaN   

                        Category Salary per annum  \
6659   Accounting & Finance Jobs            35004   
9568                     IT Jobs            62004   
11043           Engineering Jobs            39360   
12473                    IT Jobs            76908   
19142  Healthcare & Nursing Jobs            42504   
24206                    IT Jobs            45000   
24297  Healthcare & Nursing Jobs            24960   
25039           Engineering Jobs            24996   

                        SourceName         OpenDate        CloseDate  \
6659                    hays.co.uk  20130404T000000  20130703T000000   
9568                 totaljobs.com  20120205T000000  20120306T000000   
11043  jobs.planningresource.co.uk  20131006T000000  20131105T000000   
12473                 cwjobs.co.uk  20131023T150000  20131222T150000   
19142               staffnurse.com  20120505T150000  20120803T150000   
24206                totaljobs.com  20130324T150000  20130622T150000   
24297               staffnurse.com  20121018T000000  20130116T000000   
25039                  fish4.co.uk  20130608T120000  20130708T120000   

          OpenDate_parsed    CloseDate_parsed duration  
6659  2013-07-03 00:00:00 2013-04-04 00:00:00 -90 days  
9568  2012-03-06 00:00:00 2012-02-05 00:00:00 -30 days  
11043 2013-11-05 00:00:00 2013-10-06 00:00:00 -30 days  
12473 2013-12-22 15:00:00 2013-10-23 15:00:00 -60 days  
19142 2012-08-03 15:00:00 2012-05-05 15:00:00 -90 days  
24206 2013-06-22 15:00:00 2013-03-24 15:00:00 -90 days  
24297 2013-01-16 00:00:00 2012-10-18 00:00:00 -90 days  
25039 2013-07-08 12:00:00 2013-06-08 12:00:00 -30 days

Now, we can drop the columns we create before.

In [126]:
dataset1.drop(columns=['OpenDate_parsed', 'CloseDate_parsed','duration'],inplace=True) # drop columns created

In [127]:
dataset1

Id                                              Title  \
0      12612628                        Engineering Systems Analyst   
1      12612830                            Stress Engineer Glasgow   
2      12612844                   Modelling and simulation analyst   
3      12613049  Engineering Systems Analyst / Mathematical Mod...   
4      12613647         Pioneer, Miser Engineering Systems Analyst   
5      19047429            Trainee Mortgage Advisor  East Midlands   
6      20199757                   PROJECT ENGINEER, PHARMACEUTICAL   
7      20797143  Chef de Partie  Award Winning Restaurant  Exce...   
8      22579462                                   Quality Engineer   
9      22933091  Chef de Partie  Award Winning Dining  Live In ...   
10     23528672       Senior Fatigue and Damage Tolerance Engineer   
11     23529949                                C I Design Engineer   
12     23530231                            Lead Engineers (Stress)   
13     24104334    Relief Chef de Partie  Croydon, Surrey  Live in   
14     24835524        Senior Control and Instrumentation Engineer   
15     24835548               Control and Instrumentation Engineer   
16     24854800                          Electrical / ICA Engineer   
17     25337172  Pastry Chef for **** red star **** rosette hot...   
18     25452606  CHEF DE PARTIE POSITION IN **** ROSETTE HOTEL ...   
19     25452680  Senior Sous Chef for **** rosette kitchen, up ...   
20     25556432  General Manager  Funky, Cool Restaurant Concep...   
21     25892031                   MICE Sales and Marketing Manager   
22     27527047                                    C/C++ Developer   
23     27527050                               Senior PHP Developer   
24     27527063                            Senior Website Designer   
25     27527077                       Business Development Manager   
26     27739522  Welwyn Chef de Partie does it get any better t...   
27     27740574       Pastry Chef AL**** ****AA Rosette Restaurant   
28     27754996                                        QA Engineer   
29     28059808                             Documentation Engineer   
...         ...                                                ...   
25047  69228656                           Validation Engineer: IOT   
25048  69228674                     RF Design and Customer Support   
25049  69228685          Technical Sales Engineer  Deburring tools   
25050  69228842                             Senior Internal Audior   
25051  69228969  Business Development, P/T F/T, Flexible hours,...   
25052  69229164                         Account Manager (No Sales)   
25053  69229283                   Design and Applications Engineer   
25054  69229354  Part Time/Full Time Business Development Work ...   
25055  69229412       Financial Controller  temporary to permanent   
25056  69229439  ICT Trainer (Vocational Training / IT Assessor...   
25057  69229454                     Education and Training Manager   
25058  69229665                                   Storage Engineer   
25059  69229697                           Principle Java Developer   
25060  69229714                              Analytical Consultant   
25061  69229773          System Administrator Windows Exchange SQL   
25062  69229778         Junior Drainage Engineer PDS Microdrainage   
25063  69229815                       Head of Research Development   
25064  69229843        Performance Test Consultant  SoapUI, LoadUI   
25065  69229860       Rational Performance Tester  Test Consultant   
25066  69229954  Fibre Laser Physicists / Engineers (Graduate t...   
25067  69230008                      PCV Technician / Bus Mechanic   
25068  69230029                             Relationship Executive   
25069  69230036                          Cosmetic Regional Manager   
25070  69230142                      Retail Bank Modelling Manager   
25071  69230323  Business Analyst, Retail Ecommerce, Reading, B...   
25072  69245044        

## 4. Save the data

Save the data to csv file and using the default encoding

In [128]:
dataset1.to_csv('dataset1_solution.csv',encoding='utf-8',index=False)

## 5. References

[pandas](http://pandas.pydata.org/pandas-docs/stable/tutorials.html)
- - -
[Reading Materials from moodle](https://moodle.vle.monash.edu/course/view.php?id=42906&section=9#9)
- - -
[Geocoding API](https://developers.google.com/maps/documentation/geocoding/start)
- - -
[numpy](https://docs.scipy.org/doc/numpy/user/quickstart.html)
- - -
[python doc.](https://docs.python.org/3/library/index.html)